Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp

```diff
@@ -29,118 +29,300 @@
+    // preventing unnecessary splitting of layers that are spatially distant.
+    // This optimization matters because the TextureMapper uses stencil operations to render split layers,
+    // and on some hardware, stencil usage is slow. Reducing unnecessary splits helps improve performance.
+    auto potentialIntersectionsXAxis = sweepAndPrune(layers, SweepAndPruneAxis::XAxis);
+    auto potentialIntersectionsYAxis = sweepAndPrune(layers, SweepAndPruneAxis::YAxis);
+    auto potentialIntersectionsZAxis = sweepAndPrune(layers, SweepAndPruneAxis::ZAxis);
+
+    // Filter intersections and take pairs that are present in all sweep and prune sets
+    SweepAndPrunePairs potentialIntersectionsXY;
+    std::set_intersection(
+        potentialIntersectionsXAxis.begin(), potentialIntersectionsXAxis.end(),
+        potentialIntersectionsYAxis.begin(), potentialIntersectionsYAxis.end(),
+        std::inserter(potentialIntersectionsXY, potentialIntersectionsXY.begin())
+    );
+    SweepAndPrunePairs potentialIntersections;
+    std::set_intersection(
+        potentialIntersectionsXY.begin(), potentialIntersectionsXY.end(),
+        potentialIntersectionsZAxis.begin(), potentialIntersectionsZAxis.end(),
+        std::inserter(potentialIntersections, potentialIntersections.begin())
+    );
+
+    // Determine if any layer pairs intersect on the ZY plane. An intersection implies that the rendering
+    // order is either ambiguous or overlapping, necessitating the use of a BSP tree for proper ordering.
+    bool hasIntersections = false;
+    if (potentialIntersections.size() > 0) {
+        for (const auto& p : potentialIntersections) {
+            auto quadA = projectLayerToZYPlane(layers[p.first]);
+            auto quadB = projectLayerToZYPlane(layers[p.second]);
+            if (quadsIntersect(quadA, quadB)) {
+                hasIntersections = true;
+                break;
             }
         }
-    });
-
-    unsigned clipBufferSize = clipVertices.size() * sizeof(float);
-    auto clipBuffer = textureMapper.acquireBufferFromPool(clipBufferSize, TextureMapperGPUBuffer::Type::Vertex);
-    clipBuffer->updateData(clipVertices.data(), 0, clipBufferSize);
+    }
 
-    // Paint
-    traverseTree(*root, [&clipVertices, &clipBuffer, &paintLayerFunction](TextureMapperLayerNode& node) {
-        for (auto& polygon : node.polygons) {
-            unsigned numberOfClipVertices = polygon.isSplitted ? polygon.geometry.numberOfVertices() : 0;
+    // If no intersections are detected, the BSP tree building process is skipped and a fast path is taken.
+    // Other scenarios are not currently considered for optimization.
+    if (!hasIntersections) {
+        // Sort back to front
+        std::sort(layers.begin(), layers.end(), [](const Layer& layerA, const Layer& layerB) {
+            auto centerA = layerA.geometry.bounds().center();
+            auto centerB = layerB.geometry.bounds().center();
```

Comment:
I think center is not the correct way to order layers.
For example, in the diagram, layer A, B, C are drawn by sorted z order.
But, they should be drawn in the oder layer B, A, C.

<img width="702" alt="{BE5E2E4D-82D8-4FB3-BE42-FEA6E906E0EF}" src="https://github.com/user-attachments/assets/c61fef84-ab01-4d8c-a120-31913ccc0f7b" />

Reply:
So Y axis is pointing towards viewer here so projections on zy will intersect, right? In that case, this goes to BSP.

Reply:
You are right. Why do you compare centers here? I think comparing min Z is enough. Right?

Reply:
you are right, comparing to min Z should be fine

Reply:
Then, `layerA.geometry.bounds().center()` → `layerA.geometry.bounds().min().z()`